drivers/hmis/app/graphql/mutations/update_form_definition.rb

```diff
@@ -19,6 +19,8 @@ def resolve(id:, input:)
       raise 'only allowed to modify draft forms' unless definition.draft?
       raise 'not allowed to change identifier' if input.identifier.present? && input.identifier != definition.identifier
 
+      # TODO: enforce that you can't change the rule fields?
+
```

Comment:
I can't think of an elegant way to do this. My only thought is to recur on the submitted form, comparing each `link_id =>rule` and make sure nothing has changed.  Or I could "overlay" the rules from the "old" form onto the submitted one (again, based on link id).

Reply:
It seems like we could just modify `recursively_transform()` so that it overwrites the `rules` property since we are walking the tree anyway. However what happens if an element is removed by the user and then added back (or some other weirdness). Feels like in that case we'd want to pull the rules from a static source instead of the previously persisted form definition. I guess you'd want a lookup for [form_role, link_id] => {rules...}. This is sort of a combo of options 3 and 4 in 6164

Reply:
Here is an example where we move the "Static source" from JSON files into code: https://github.com/greenriver/hmis-warehouse/pull/4478/files -- is that sort of what you had in mind?

I guess the other point here is that we could rely on the frontend to ensure that `rule`s aren't changed. (It should not provide an interface for editing them). Obviously someone could change them with the API in that case, but that's not a dangerous operation, just not ideal for HUD Compliance. So if we're fine with that, we could just this as-is.

The other field that's important is `data_collected_about`. That field can be changed by the user but only expanded beyond the HUD required "minimum", so we still need a mechanism for that. The linked branch has an approach, but there are other ways we could do it.

Reply:
> Here is an example where we move the "Static source" from JSON files into code: https://github.com/greenriver/hmis-warehouse/pull/4478/files -- is that sort of what you had in mind?

Yes, that's what I was thinking.
> I guess the other point here is that we could rely on the frontend to ensure that rules aren't changed.

The goal is to prevent folks from accidentally getting out of compliance when customizing their forms right? I could see an approach where we assume the front-end will always preserve the hidden "hud_rules". If an admin removes required elements or there's a client bug that clobbers parts of the form, then we could just decide not to handle that since it is unlikely.

But I don't quite understand the consequences of someone falling out of compliance. I'm guessing it could be kind of a big problem particularly if nobody catches it for a while. It seems like we could ward this off with a combination of validation and forcing to known-good values:
* validate that the appropriate fields are present in the form given the role
* validate or force hud_rules and data_collected_about to known-good values (maybe others)

If this feels tricky to do on every save, maybe it could be something that occurs during publication. For example you can't publish a form unless it passes a hud-compliance check

Reply:
Yeah I agree with that -- it could be a problem if non-compliant for a while. I think it makes sense to ensure compliance on publish. I'll move forward with that.